docs/design/0012-generic-association.md

```diff
@@ -0,0 +1,275 @@
+ * `elasticsearch-host`: hostname of the Elasticsearch cluster
+ * `elasticsearch-port`: port of the Elasticsearch cluster
+ * `elasticsearch-ca`: Certificate of the Elasticsearch CA
+ * `kibana-host`: hostname to be used to access Kibana
+ * `kibana-port`: port of Kibana
+ * `kibana-ca`: certificate of the Kibana CA
+ * `apmserver-url`: url to be used to access the APM Server
+ * `apmserver-token`: token to be used to access the APM Server
+ * `apmserver-ca`: certificate of the APM Server CA
+
+## Injecting association data automatically
+
+#### Option #1: Updating the Pod's parent resource
+
+Two different approaches might be used in order to populate the environment variables and bind the `Secret` automatically in a `Pod`.
+
+In the first one ECK would insert some environment variables in the `Pod`'s parent resource:
+
+Here is the `Deployment` as it would be created by a user:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-application
+spec:
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: container1
+        image: ...
+      - name: container2
+        image: ...
+```
+
+Here is the `Deployment` as it would be updated by ECK:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-application
+spec:
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      annotations:
+        elastic.config.checksum: 585696 # allows an automatic rotation
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: container1
+        image: ...
+        env:
+          - name: ELASTICSEARCH_USERNAME
+            valueFrom:
+              secretKeyRef:
+                name: <automatic-generated-name>
+                key: username
+        volumeMounts:
+        - name: elastic-volume
+          mountPath: /mnt/elastic
+      - name: container2
+        image: ...
+        volumeMounts:
+        - name: elastic-volume
+          mountPath: /mnt/elastic
+      volumes:
+      - name: elastic-volume
+        secret:
+          secretName: <automatic-generated-name>
+```
+
+  * Pros:
+    * The configuration checksum can be injected in the pod template to allow an automatic restart of the `Pods` 
+    * ECK can inject and manage some "well-known" environment variables, some good examples are the ones used in the context of the APM Server (`ELASTIC_APM_SERVER_URLS`, `ELASTIC_APM_SECRET_TOKEN`...) 
+  * Cons:
+    * It could be considered as "intrusive"
+    * It could be a lot of resources to watch
+
+#### Option #2: Using a MutatingAdmissionWebhook
+
+  * Pros:
+    * Less intrusive than updating the parent resource
+  * Cons:
+    * Cannot handle the lifecycle of the Secret content. The associated application must detect than something has changed in the `Secret`. In case of an environment variable the application must handle a `Pod` restart itself.
```

Comment:
Would it make sense to optionally (opt-in annotation) inject a liveness probe that would fail if the secret gets updated? Based on an additional "checksum" entry in the secret maybe.